fillers/eips/eip4844/datahash_opcode.py

```diff
@@ -538,118 +545,238 @@ def test_datahash_gas(_: str):
-        ),
     }
+    post = {}
+
+    def create_tx(ty: int, to: str, nonce: int):
+        return Transaction(
+            ty=ty,
+            nonce=nonce,
+            data=to_hash_bytes(0),
+            to=to,
+            gas_limit=3000000,
+            max_fee_per_gas=10,
+            max_priority_fee_per_gas=10,
+            max_fee_per_data_gas=10,
+            access_list=[],
+            blob_versioned_hashes=[to_hash_bytes(2**256 - 1)]
+            if ty == 5
+            else None,
+        )
+
+    txs_ty2, txs_ty5 = [], []
+    for i, code_gas_measure in enumerate(gas_measures_code):
+        address = to_address(0x100 + i * 0x100)
+        pre[address] = Account(code=code_gas_measure)
+        txs_ty2.append(create_tx(ty=2, to=address, nonce=i))
+        txs_ty5.append(create_tx(ty=5, to=address, nonce=i))
+        post[address] = Account(storage={0: DATAHASH_GAS_COST})
+
+    # DATAHASH gas cost on tx type 2
+    yield StateTest(
+        env=env,
+        pre=pre,
+        post=post,
+        txs=txs_ty2,
+        tag="tx_type_2",
+    )
 
     # DATAHASH gas cost on tx type 5
-    tx1 = Transaction(
-        ty=5,
-        nonce=0,
-        data=to_hash_bytes(0),
-        to=datahash_zero_gas_measure_code_address,
-        gas_limit=3000000,
-        max_fee_per_gas=10,
-        max_priority_fee_per_gas=10,
-        max_fee_per_data_gas=10,
-        access_list=[],
-        blob_versioned_hashes=[to_hash_bytes(2**256 - 1)],
+    # Can only use 4 txs (max blobs) per block
+    num_blocks = (len(txs_ty5) + MAX_BLOB_PER_BLOCK - 1) // MAX_BLOB_PER_BLOCK
+    blocks = [
+        Block(
+            txs=txs_ty5[i * MAX_BLOB_PER_BLOCK : (i + 1) * MAX_BLOB_PER_BLOCK]
+        )
+        for i in range(num_blocks)
+    ]
+
+    yield BlockchainTest(
+        pre=pre,
+        post=post,
+        blocks=blocks,
+        tag="tx_type_5",
     )
-    tx2 = Transaction(
-        ty=5,
-        nonce=1,
-        data=to_hash_bytes(0),
-        to=datahash_max_gas_measure_code_address,
-        gas_limit=3000000,
-        max_fee_per_gas=10,
-        max_priority_fee_per_gas=10,
-        max_fee_per_data_gas=10,
-        access_list=[],
-        blob_versioned_hashes=[to_hash_bytes(2**256 - 1)],
+
+
+@test_from(fork=ShardingFork)
+def test_datahash_blob_versioned_hash(_: str):
+    """
+    Tests that the `DATAHASH` opcode returns the correct versioned hash for
+    various valid indexes
+    """
+    NUM_BLOCKS = 10
+
+    pre = {
+        TestAddress: Account(balance=10000000000000000000000),
+    }
+    blocks = []
+    post = {}
+
+    datahash_verbatim = "verbatim_{}i_{}o".format(
+        Op.DATAHASH.popped_stack_items, Op.DATAHASH.pushed_stack_items
     )
 
-    post = {
-        datahash_zero_gas_measure_code_address: Account(
-            storage={
-                0: DATAHASH_GAS_COST,
-            }
-        ),
-        datahash_max_gas_measure_code_address: Account(
+    datahash_code = Yul(
```

Comment:
I think we should generate this bytecode using `Op` and a python loop instead of relying on Yul.